examples/nextjs-with-typescript/pages/_document.tsx

```diff
@@ -27,57 +26,6 @@ export default class MyDocument extends Document {
 
 // `getInitialProps` belongs to `_document` (instead of `_app`),
 // it's compatible with static-site generation (SSG).
-MyDocument.getInitialProps = async (ctx) => {
-  // Resolution order
-  //
-  // On the server:
-  // 1. app.getInitialProps
-  // 2. page.getInitialProps
-  // 3. document.getInitialProps
-  // 4. app.render
-  // 5. page.render
-  // 6. document.render
-  //
-  // On the server with error:
-  // 1. document.getInitialProps
-  // 2. app.render
-  // 3. page.render
-  // 4. document.render
-  //
-  // On the client
-  // 1. app.getInitialProps
-  // 2. page.getInitialProps
-  // 3. app.render
-  // 4. page.render
-
-  const originalRenderPage = ctx.renderPage;
-
-  // You can consider sharing the same emotion cache between all the SSR requests to speed up performance.
-  // However, be aware that it can have global side effects.
-  const cache = createEmotionCache();
-  const { extractCriticalToChunks } = createEmotionServer(cache);
-
-  ctx.renderPage = () =>
-    originalRenderPage({
-      enhanceApp: (App: any) => (props) => <App emotionCache={cache} {...props} />,
-    });
-
-  const initialProps = await Document.getInitialProps(ctx);
-  // This is important. It prevents emotion to render invalid HTML.
-  // See https://github.com/mui-org/material-ui/issues/26561#issuecomment-855286153
-  const emotionStyles = extractCriticalToChunks(initialProps.html);
-  const emotionStyleTags = emotionStyles.styles.map((style) => (
-    <style
-      data-emotion={`${style.key} ${style.ids.join(' ')}`}
-      key={style.key}
-      // eslint-disable-next-line react/no-danger
-      dangerouslySetInnerHTML={{ __html: style.css }}
-    />
-  ));
-
-  return {
-    ...initialProps,
-    // Styles fragment is rendered after the app and page rendering finish.
-    styles: [...React.Children.toArray(initialProps.styles), ...emotionStyleTags],
-  };
-};
+MyDocument.getInitialProps = withMuiSetup({
```

Comment:
I think that the objective should be to import directly from `@mui/core-nextjs` and here it feels too opinionated. I don't see how developers will be able to extend getInitialProps. IMHO, for this method, more boilerplate is better, until Next.js has a property extendable plugin system.

Reply:
Right, we could maintain some `@mui/core-nextjs` and be precise in explaining the intention.

TThe reason I went with the closure is that I am combining it with another function that keeps injecting more things (but yeah ... hopefully NextJS fixes the issue)

Reply:
I share Olivier's worry, creating a new abstraction is worse rather than better. People struggle even now to configure nextjs properly. If they need to extend the `getInitialProps` now they have to learn another API. 

It is way more cleaner, and it makes total sense to me, but for new comers it may be overwhelming